DRILL-8056: Add OAuth2 Support for HTTP Rest Plugin

cgivre opened the pull request:
# [DRILL-8056](https://issues.apache.org/jira/browse/DRILL-8056): Add OAuth2 Support for HTTP Rest Plugin

## Description
This PR adds the ability to query APIs which use OAuth2.0.  The PR handles obtaining all the tokens and refreshing them in API calls. 

## Documentation

https://github.com/apache/drill/blob/d82e039c6af4d6996d83ce6619c599c90ced8637/contrib/storage-http/OAuth.md

See link: 👆

### Tasks Remaining

- [x] Complete Documentation

## Testing
Manually tested and added unit tests.

vvysotskyi commented:
Looks like there are some issues with added tests:
```
Error:  org.apache.drill.exec.store.http.oauth.TestOAuthAccessTokenRepository.testGetAccessToken  Time elapsed: 0.042 s  <<< ERROR!
java.net.BindException: Address already in use (Bind failed)
at org.apache.drill.exec.store.http.oauth.TestOAuthAccessTokenRepository.testGetAccessToken(TestOAuthAccessTokenRepository.java:56)
```

cgivre commented:
> Looks like there are some issues with added tests:
> 
> ```
> Error:  org.apache.drill.exec.store.http.oauth.TestOAuthAccessTokenRepository.testGetAccessToken  Time elapsed: 0.042 s  <<< ERROR!
> java.net.BindException: Address already in use (Bind failed)
> at org.apache.drill.exec.store.http.oauth.TestOAuthAccessTokenRepository.testGetAccessToken(TestOAuthAccessTokenRepository.java:56)
> ```

That was weird.  The test ran fine in my IDE.   I think I fixed it now.

vvysotskyi commented on exec/jdbc-all/pom.xml:
```diff
@@ -562,7 +562,7 @@
 
                   </message>
-                  <maxsize>46600000</maxsize>
+                  <maxsize>50000000</maxsize>
```
Is there any reason for updating this limit? The functionality is only for HTTP plugin, so it shouldn't affect jdbc driver size.

cgivre replied:
I'm not sure why this happened.  I did bump the version of `okhttp3` to 4.9.1 or whatever the latest version is, but there are no new dependencies.

vvysotskyi replied:
Before your change, okhttp3 was present only for test scope (and used only in some unit tests), and therefore it wasn't included in the JDBC driver. But you have changed its scope from the test to compile, so it is added to the driver. But the question here: is it required in JDBC?

cgivre replied:
@vvysotskyi Thanks for looking at this.  It actually is used in one of the files for Drill's internal rest API.  (See below). Does that need to be included in the JDBC driver?

https://github.com/apache/drill/blob/d6c54936129f8e2a8eda83be7b1ab69998e893f8/exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/StorageResources.java#L206-L12

vvysotskyi replied:
I don't think it should be included, so please exclude it from the driver.

cgivre replied:
@vvysotskyi Fixed!

cgivre replied:
@vvysotskyi Unfortunately I tried excluding it from the JDBC and it does need to be there.

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/http/HttpOAuthConfig.java:
```diff
@@ -0,0 +1,173 @@
+
+  @JsonPOJOBuilder(withPrefix = "")
+  public static class HttpOAuthConfigBuilder {
+    @Getter
+    @Setter
```
Instead of adding annotations for every field, you can add them to the class and they will be applied to every field.

cgivre replied:
Fixed.

vvysotskyi commented on contrib/storage-http/src/test/java/org/apache/drill/exec/store/http/oauth/TestOAuthAccessTokenRepository.java:
```diff
@@ -0,0 +1,168 @@
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+
+public class TestOAuthAccessTokenRepository {
```
Please extend the test class from `BaseTest`.

cgivre replied:
I actually removed this entire class.  Added different tests that better test the whole process.

vvysotskyi commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/oauth/AccessTokenAuthenticator.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  @Override
+  public Request authenticate(Route route, @NotNull Response response) {
```
According to this logic, the access token will be regenerated for every `authenticate` call. So do we actually need to store it in the storage config and persist it to the zookeeper instead of holding it for a specific HTTP client?

cgivre replied:
The way it should work is that it first tries the request with the access token that is stored in Drill.  If that request fails, it executes additional requests to obtain an updated token, and then re-executes the request.

vvysotskyi commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/oauth/AccessTokenRepository.java:
```diff
@@ -0,0 +1,149 @@
+
+    // This null check is here for testing only.  In actual Drill, the registry will not be null.
+    if (registry != null) {
+      OAuthUtils.updateOAuthTokens(registry, updatedConfig, pluginConfig);
```
Please see the previous comment regarding storing access tokens in ZK...

cgivre replied:
This is completely refactored to store (and update) the creds in the `credentialProvider`

vvysotskyi reviewed (changes requested):
@cgivre, thanks for the PR, it will be a significant improvement for the HTTP storage plugin.

I have several objections regarding adding HTTP plugin-specific code to the java-exec and continuously updating the storage plugin registry with the storage plugin tokens.

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/StorageResources.java:
```diff
@@ -180,6 +193,48 @@ public Response enablePlugin(@PathParam("name") String name, @PathParam("val") B
+  public Response updateAuthToken(@PathParam("name") String name, @QueryParam("code") String code) {
+    try {
+      if (storage.getPlugin(name).getConfig().getClass().getSimpleName().equalsIgnoreCase("HttpStoragePluginConfig")) {
+        HttpStoragePluginConfig config = (HttpStoragePluginConfig)storage.getPlugin(name).getConfig();
```
I refactored this so that it uses the `credentialProvider` to store the credentials.

jnturton commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/security/CredentialProviderUtils.java:
```diff
@@ -43,4 +43,59 @@ public static CredentialsProvider getCredentialsProvider(
   }
+
+  /**
+   * Constructor for OAuth based authentication.  Allows for tokens to be stored in whatever vault
```
I think we can avoid having all of this OAuth-specific code show up here like this, by a universally-supported `setCredential()`

cgivre commented on exec/jdbc-all/pom.xml:
```diff
@@ -562,7 +562,7 @@
 
                   </message>
-                  <maxsize>46600000</maxsize>
+                  <maxsize>50000000</maxsize>
```
Unfortunately, I had to include them because there is an `OAuthUtils` class which uses OkHttp which is part of java-exec.  If you have some advice as to where else I could put that, I'm all ears.

The issue is that both the rest server and the http storage plugin needs to be able to access the same body of code for obtaining and refreshing tokens.

jnturton replied:
@cgivre I _think_ you include in java-exec's pom and also explcitly _exclude_ in the java-exec dependency section in jdbc-all's pom.

cgivre replied:
The problem is that it is needed for `java-exec`.  I tried this approach and unfortunately, it didn't work.

cgivre commented on contrib/storage-http/OAuth.md:
```diff
@@ -0,0 +1,122 @@
+* `accessToken`: The token which is used to grant access to a protected resource
+* `refreshToken`: The token used to obtain a new `accessToken` without having to re-authorize the application.
+
+Currently, Drill does not allow per-user credentials.  However, it is possible to configure Drill to use the Vault Credential Provider so that each individual user has their 
```
You are correct... :-)

cgivre commented on logical/src/main/java/org/apache/drill/common/logical/security/CredentialsProvider.java:
```diff
@@ -33,6 +37,16 @@
   @JsonIgnore
   Map<String, String> getCredentials();
+
+  @JsonIgnore
+  default void updateCredentials(String key, String value) throws UserException {
+    throw UserException.internalError()
+      .message("Update credential function not implemented for " + Id.NAME)
+      .build(logger);
+  }
```
Fixed

cgivre commented on exec/java-exec/src/main/resources/rest/storage/update.ftl:
```diff
@@ -133,6 +138,71 @@
       }
     });
 
+    $("#getOauth").click(function() {
```
Fixed.

cgivre commented:
@vvysotskyi  I added a parameter to put the access token in the URL instead of the header.  This is technically not part of the OAuth spec....

vvysotskyi reviewed (commented):
Thanks for addressing CR comments and introducing additional improvements for this feature!

I thought about the way how it can be refactored, and one of the ideas I have is to move all the logic related to obtaining and regenerating access_tocken to the OAuth credentials provider. So it will be also responsible for detecting whether the token is outdated, regenerating it, and updating it in the registry instead of doing that.

One of the unresolved issues I see there are concurrent updates of the access_token. Let's assume we have a cluster of two drillbits, one outdated token, and queries that will be running concurrently. So, in this case, it is possible that the first drillbit will regenerate the token, after that the second one will regenerate the token, store it in the persistent store and after that, the first drillbit will store its token that will be actually outdated because the previous one was regenerated. One of the solutions is to use distributed lock similar to the way how it is done for dynamic UDFs feature or so on.

Also, if we will provide a separate persistent store for credentials, we will be able to update it only instead of updating the whole plugin configuration.

cgivre commented:
@vvysotskyi 
To answer your questions, I actually first attempted this with a dedicated `OAuthCredentialProvider` class.  The issue I ran into was that it needed the `SimpleHttp` class within the `HttpStoragePlugin` to actually do all the REST calls.  For instance, if a user has a certificate store or web proxy configured, that would all be done in the config for the HTTP plugin and thus, you'd need that when making the token refreshes. 

I was thinking about trying to make this reusable for other plugins that may use OAuth.  The issue there is that while the flow is the same, there are a lot of pieces which are different.  For example, I've been poking at a  storage plugin for Google Sheets, which uses OAuth. While both use OAuth, the configuration is completely different and it unfortunately isn't really possible to reuse much in that situation. 

The other thing to consider is that right now, the OAuth work only uses the `PlainCredentialProvider`, however users might want to store these tokens in Vault.  I was thinking as a v2, to extend the functionality so that the tokens can be stored using the `VaultCredentialProvider`, but that will have to wait for another PR.  Since some of these tokens are sensitive, it actually would make a lot of sense to store them more securely than in plain text in Zookeeper. ;-)

Regarding concurrency, let's say that two users execute a query at virtually the same time.  Let's say that `access_token_A` is expired.  What should happen is that the first query to hit that server will get a 401 error.  Drill would then send a new request with the `refresh_token` and update the access token to `access_token_B`.   My understanding of OAuth is that the tokens time out, but during that time, if userB requested a new access token using the same refresh_token, userB would get the same updated access token (`access_token_b`).  In this situation, I think the worst case scenario would be that a few extra refresh requests are sent.

I think I may have mentioned this, but I'm also working on extending this a bit to allow for per-user tokens and credentials, so when that gets implemented, the chances of collisions happening is even less.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpOAuthConfig.java:
```diff
@@ -0,0 +1,86 @@
+@EqualsAndHashCode
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
+@JsonDeserialize(builder = HttpOAuthConfig.HttpOAuthConfigBuilder.class)
+public class HttpOAuthConfig {
```
The `OAuthConfig` class is used to hold several variables that don't end up in the credential provider.   Specifically, these are config parameters like `accessTokenInHeader`, `scope`, `tokenType` etc.   These didn't really seem like they belong in the `credentialsProvider`.

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/StorageResources.java:
```diff
@@ -180,6 +194,47 @@ public Response enablePlugin(@PathParam("name") String name, @PathParam("val") B
+  @Produces(MediaType.TEXT_HTML)
+  public Response updateAuthToken(@PathParam("name") String name, @QueryParam("code") String code) {
+    try {
+      if (storage.getPlugin(name).getConfig().getClass().getSimpleName().equalsIgnoreCase("HttpStoragePluginConfig")) {
```
Fixed

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/oauth/AccessTokenAuthenticator.java:
```diff
@@ -0,0 +1,71 @@
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.Route;
+import org.jetbrains.annotations.NotNull;
```
Fixed, I think.

vvysotskyi commented:
@cgivre, for now, we can query Dropbox using the file storage plugin, but Dropbox also supports OAuth (in their docs I didn't find differences with OAuth), so extracting that logic to the OAuth creds provider and removing dependency to HTTP plugin will automatically enable this and other similar cases. If we have some specific logic for Google Sheets, in the future we can extend that implementation of the OAuth creds provider to handle additional cases.

Regarding concurrent updates of the token, I think it is a major issue because we could have high concurrent queries (even for a single user) that will be slowed down because of unnecessary updating of the access_token. Please take a look at the versioned persistent store, I think it can be used to solve that issue.

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/StorageResources.java:
```diff
@@ -62,6 +72,10 @@
 @RolesAllowed(ADMIN_ROLE)
 public class StorageResources {
   private static final Logger logger = LoggerFactory.getLogger(StorageResources.class);
+  private static final String OAUTH_SUCCESS = "<html>\n" + "<head>\n" + "  <title>Success!</title>\n" + "  <link rel=\"shortcut icon\" href=\"/static/img/drill.ico\">\n" + "  " +
```
Fixed.

cgivre commented:
@vvysotskyi Thanks for your review comments.  I believe I addressed all your concerns.  The ephemeral tokens are now stored in a persistent store, with the `client_id` and `client_secret` being stored using the `credential provider`. 
Thanks!

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/oauth/Tokens.java:
```diff
@@ -0,0 +1,73 @@
+  /**
+   * Associates provided token with provided plugin in token table.
+   *
+   * @param token   alias of the value to associate with
+   * @param value   value that will be associated with provided alias
+   * @param replace whether existing value for the same alias should be replaced
+   * @return {@code true} if provided alias was associated with
+   * the provided value in aliases table
```
And here

cgivre replied:
Fixed.

vvysotskyi commented on logical/src/main/java/org/apache/drill/common/logical/security/PlainCredentialsProvider.java:
```diff
@@ -47,6 +51,18 @@ public PlainCredentialsProvider(@JsonProperty("credentials") Map<String, String>
 
+  @Override
+  @JsonIgnore
+  public void setCredential(String key, String value) {
```
Please delete this method as it is not used anywhere.

cgivre replied:
Removed.

vvysotskyi commented on logical/src/main/java/org/apache/drill/common/logical/security/CredentialsProvider.java:
```diff
@@ -33,6 +37,20 @@
+   * value to persistent storage.
+   */
+  @JsonIgnore
+  default void setCredential(String key, String value) throws UserException {
```
And please remove this one.

vvysotskyi commented on pom.xml:
```diff
@@ -467,6 +467,7 @@
             <exclude>**/*.dbf</exclude>
             <exclude>**/*.cnf</exclude>
             <exclude>**/*.access_log</exclude>
+            <exclude>**/success.ftl</exclude>
```
Please fix the license header instead of excluding the file.

cgivre replied:
Done.  I renamed the file `success.html`.

vvysotskyi reviewed (approved):
@cgivre, thanks for making more changes, I've added several comments, please address them, and merge this PR after that.

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/oauth/TokenRegistry.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Map;
+
+/**
+ * Registry for public and user-owned aliases.
```
Fixed

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/oauth/Tokens.java:
```diff
@@ -0,0 +1,73 @@
+package org.apache.drill.exec.oauth;
+
+/**
+ * Aliases table. Provides API for managing and obtaining aliases.
```
Fixed.  I removed the Javadoc for this.
